ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -4,58 +4,202 @@ import {
+  ctx.lineTo(arrivalTime + ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime + ARROW_WIDTH, OCCUPANCY_ZONE_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTime - ARROW_WIDTH, OCCUPANCY_ZONE_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.fill();
+  ctx.moveTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime + ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.stroke();
+};
+
+type DrawRemainingTrainsBox = {
+  ctx: CanvasRenderingContext2D;
+  text: string;
+  textX: number;
+  textY: number;
+};
+
+const drawRemainingTrainsBox = ({ ctx, text, textX, textY }: DrawRemainingTrainsBox) => {
+  ctx.fillStyle = REMAINING_TRAINS_BACKGROUND;
+  ctx.beginPath();
+  ctx.rect(textX, textY, REMAINING_TRAINS_WIDTH, REMAINING_TRAINS_HEIGHT);
+  ctx.fill();
+  ctx.stroke();
+  ctx.fillStyle = WHITE_100;
+  ctx.font = SANS;
+  ctx.textAlign = 'center';
+  ctx.textBaseline = 'middle';
+  ctx.fillText(text, textX + REMAINING_TRAINS_WIDTH / 2, textY + REMAINING_TRAINS_HEIGHT / 2);
+};
+
 export const drawOccupancyZones = ({
   ctx,
   width,
   height,
   tracks,
   occupancyZones,
   getTimePixel,
+  selectedTrainId,
 }: {
   ctx: CanvasRenderingContext2D;
   width: number;
   height: number;
   tracks: Track[] | undefined;
   occupancyZones: OccupancyZone[] | undefined;
   getTimePixel: (time: number) => number;
+  selectedTrainId: string;
 }) => {
   ctx.clearRect(0, 0, width, height);
 
   tracks?.forEach((track, index) => {
     const trackTranslate = index === 0 ? CANVAS_PADDING : TRACK_HEIGHT_CONTAINER;
     ctx.translate(0, trackTranslate);
 
-    const trackOccupancyZones = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const filteredOccupancyZone = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const sortedOccupancyZone = filteredOccupancyZone?.sort(
+      (a, b) => a.arrivalTime.getTime() - b.arrivalTime.getTime()
+    );
+
+    if (!sortedOccupancyZone) return;
 
-    if (!trackOccupancyZones) return;
+    let primaryArrivalTimePixel = 0;
+    let primaryDepartureTimePixel = 0;
+    let lastDepartureTimePixel = 0;
+    let yPosition = OCCUPANCY_ZONE_START;
+    let yOffset = 0;
+    let counter = 0;
 
-    trackOccupancyZones.forEach((zone) => {
+    if (sortedOccupancyZone && sortedOccupancyZone.length > 1) {
+      primaryArrivalTimePixel = getTimePixel(sortedOccupancyZone[0].arrivalTime.getTime());
+      primaryDepartureTimePixel = getTimePixel(sortedOccupancyZone[0].departureTime.getTime());
+      lastDepartureTimePixel = primaryDepartureTimePixel;
+    }
+
+    // use a for of loop to be able to break the loop if there are more than 9 zones
+    for (const [sortedIndex, zone] of sortedOccupancyZone.entries()) {
       const arrivalTime = getTimePixel(zone.arrivalTime.getTime());
       const departureTime = getTimePixel(zone.departureTime.getTime());
```

Comment:
Do you think you can rename these consts ? Maybe in a separated commit
```suggestion
      const arrivalTimePixel = getTimePixel(zone.arrivalTime.getTime());
      const departureTimePixel = getTimePixel(zone.departureTime.getTime());
```

Reply:
done